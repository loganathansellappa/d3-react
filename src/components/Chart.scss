.chart-container {
  display: flex;
  margin-top: 120px;
  justify-content: center;
  width: 100%;
}

/* Tooltip styles */
//.tooltip {
//  display: none;
//  position: absolute;
//  background: rgba(0, 0, 0, 0.7);
//  color: white;
//  padding: 5px;
//  border-radius: 5px;
//  font-size: 12px;
//
//  & div {
//    margin: 5px 0;
//  }
//
//  &:after {
//    content: '';
//    position: absolute;
//    border-style: solid;
//    border-width: 10px 10px 0;
//    border-color: #fff transparent transparent transparent;
//    top: 100%;
//    left: 50%;
//    transform: translateX(-50%);
//  }
//}


.tooltip {
  position: absolute;
  padding: 15px;
  background-color: rgb(11, 40, 65);
  color: white;
  border: 0px solid white;
  border-radius: 0px;
  display: none;
  opacity: .75;
  font-size: 14px;
}

.tooltip-line {
  stroke-opacity: 0.5;
}

.tooltip div {
  margin: 5px 0; /* Add some spacing between rows */
}

.tooltip:after {
  content: '';
  position: absolute;
  border-style: solid;
  border-width: 10px 10px 0;
  border-color: #fff transparent transparent transparent;
  top: 100%;
  left: 50%;
  transform: translateX(-50%);
}
