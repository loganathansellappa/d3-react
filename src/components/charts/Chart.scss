.chart-wrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  .button-container {
    align-self: flex-start;
    margin-left: 10%;
    margin-top: 10px;
  }
}

.tooltip {
  position: absolute;
  padding: 15px;
  background-color: rgb(11, 40, 65);
  color: white;
  border: 0px solid white;
  border-radius: 0px;
  display: none;
  opacity: 0.75;
  font-size: 14px;
}

.tooltip-line {
  stroke-opacity: 0.5;
}

.tooltip div {
  margin: 1px 0;
}