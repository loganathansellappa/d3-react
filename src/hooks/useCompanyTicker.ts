import { useLocation } from "react-router-dom";

export const useCompanyTicker = (): string => {
  const location = useLocation();
  const searchParams = new URLSearchParams(location.search);
  const sym = searchParams.get("sym");
  return sym || "IBM"; // Provide the default value here
};
