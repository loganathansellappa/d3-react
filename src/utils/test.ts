import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";

export const renderWithRouter = (ui: JSX.Element, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
export const mockOverviewResponse = () => {
  return {
    Symbol: "IBM",
    AssetType: "Common Stock",
    Name: "International Business Machines",
    Description:
      "International Business Machines Corporation (IBM) is an American multinational technology company headquartered in Armonk, New York, with operations in over 170 countries. The company began in 1911, founded in Endicott, New York, as the Computing-Tabulating-Recording Company (CTR) and was renamed International Business Machines in 1924. IBM is incorporated in New York. IBM produces and sells computer hardware, middleware and software, and provides hosting and consulting services in areas ranging from mainframe computers to nanotechnology. IBM is also a major research organization, holding the record for most annual U.S. patents generated by a business (as of 2020) for 28 consecutive years. Inventions by IBM include the automated teller machine (ATM), the floppy disk, the hard disk drive, the magnetic stripe card, the relational database, the SQL programming language, the UPC barcode, and dynamic random-access memory (DRAM). The IBM mainframe, exemplified by the System/360, was the dominant computing platform during the 1960s and 1970s.",
    CIK: "51143",
    Exchange: "NYSE",
    Currency: "USD",
    Country: "USA",
    Sector: "TECHNOLOGY",
    Industry: "COMPUTER & OFFICE EQUIPMENT",
    Address: "1 NEW ORCHARD ROAD, ARMONK, NY, US",
    FiscalYearEnd: "December",
    LatestQuarter: "2023-09-30",
    MarketCapitalization: "133799322000",
    EBITDA: "13663000000",
    PERatio: "18.91",
    PEGRatio: "1.276",
    BookValue: "25.28",
    DividendPerShare: "6.62",
    DividendYield: "0.0457",
    EPS: "7.75",
    RevenuePerShareTTM: "67.3",
    ProfitMargin: "0.113",
    OperatingMarginTTM: "0.145",
    ReturnOnAssetsTTM: "0.0455",
    ReturnOnEquityTTM: "0.328",
    RevenueTTM: "61170999000",
    GrossProfitTTM: "32688000000",
    DilutedEPSTTM: "7.75",
    QuarterlyEarningsGrowthYOY: "0.126",
    QuarterlyRevenueGrowthYOY: "0.046",
    AnalystTargetPrice: "137.35",
    TrailingPE: "18.91",
    ForwardPE: "15.55",
    PriceToSalesRatioTTM: "2.108",
    PriceToBookRatio: "6.75",
    EVToRevenue: "2.969",
    EVToEBITDA: "25.81",
    Beta: "0.771",
    ["52WeekHigh"]: "151.93",
    ["52WeekLow"]: "118.71",
    ["50DayMovingAverage"]: "143.62",
    ["200DayMovingAverage"]: "135.29",
    SharesOutstanding: "911006000",
    DividendDate: "2023-09-09",
    ExDividendDate: "2023-11-08",
  };
};

export const mockBalanceSheetResponse = () => {
  return {
    symbol: "IBM",
    annualReports: [
      {
        fiscalDateEnding: "2022-12-31",
        reportedCurrency: "USD",
        totalAssets: "127243000000",
        totalCurrentAssets: "29118000000",
        cashAndCashEquivalentsAtCarryingValue: "7886000000",
        cashAndShortTermInvestments: "7886000000",
        inventory: "1552000000",
        currentNetReceivables: "14209000000",
        totalNonCurrentAssets: "96874000000",
        propertyPlantEquipment: "5334000000",
        accumulatedDepreciationAmortizationPPE: "13361000000",
        intangibleAssets: "67133000000",
        intangibleAssetsExcludingGoodwill: "11184000000",
        goodwill: "55949000000",
        investments: "None",
        longTermInvestments: "142000000",
        shortTermInvestments: "852000000",
        otherCurrentAssets: "2610000000",
        otherNonCurrentAssets: "None",
        totalLiabilities: "105222000000",
        totalCurrentLiabilities: "31505000000",
        currentAccountsPayable: "4051000000",
        deferredRevenue: "15531000000",
        currentDebt: "9511000000",
        shortTermDebt: "4760000000",
        totalNonCurrentLiabilities: "83414000000",
        capitalLeaseObligations: "164000000",
        longTermDebt: "47190000000",
        currentLongTermDebt: "4676000000",
        longTermDebtNoncurrent: "46189000000",
        shortLongTermDebtTotal: "107759000000",
        otherCurrentLiabilities: "9788000000",
        otherNonCurrentLiabilities: "12243000000",
        totalShareholderEquity: "21944000000",
        treasuryStock: "169484000000",
        retainedEarnings: "149825000000",
        commonStock: "58343000000",
        commonStockSharesOutstanding: "906091977",
      },
      {
        fiscalDateEnding: "2021-12-31",
        reportedCurrency: "USD",
        totalAssets: "132001000000",
        totalCurrentAssets: "29539000000",
        cashAndCashEquivalentsAtCarryingValue: "6650000000",
        cashAndShortTermInvestments: "6650000000",
        inventory: "1649000000",
        currentNetReceivables: "14977000000",
        totalNonCurrentAssets: "101786000000",
        propertyPlantEquipment: "5694000000",
        accumulatedDepreciationAmortizationPPE: "14390000000",
        intangibleAssets: "68154000000",
        intangibleAssetsExcludingGoodwill: "12511000000",
        goodwill: "55643000000",
        investments: "199000000",
        longTermInvestments: "159000000",
        shortTermInvestments: "600000000",
        otherCurrentAssets: "5663000000",
        otherNonCurrentAssets: "None",
        totalLiabilities: "113005000000",
        totalCurrentLiabilities: "33619000000",
        currentAccountsPayable: "3955000000",
        deferredRevenue: "16095000000",
        currentDebt: "13551000000",
        shortTermDebt: "6787000000",
        totalNonCurrentLiabilities: "90188000000",
        capitalLeaseObligations: "63000000",
        longTermDebt: "56193000000",
        currentLongTermDebt: "6728000000",
        longTermDebtNoncurrent: "44917000000",
        shortLongTermDebtTotal: "110496000000",
        otherCurrentLiabilities: "9386000000",
        otherNonCurrentLiabilities: "13996000000",
        totalShareholderEquity: "18901000000",
        treasuryStock: "169392000000",
        retainedEarnings: "154209000000",
        commonStock: "57319000000",
        commonStockSharesOutstanding: "898068600",
      },
    ],
    quarterlyReports: [
      {
        fiscalDateEnding: "2023-09-30",
        reportedCurrency: "USD",
        totalAssets: "129321000000",
        totalCurrentAssets: "27705000000",
        cashAndCashEquivalentsAtCarryingValue: "7257000000",
        cashAndShortTermInvestments: "7257000000",
        inventory: "1399000000",
        currentNetReceivables: "6039000000",
        totalNonCurrentAssets: "100035000000",
        propertyPlantEquipment: "5369000000",
        accumulatedDepreciationAmortizationPPE: "12848000000",
        intangibleAssets: "70874000000",
        intangibleAssetsExcludingGoodwill: "11278000000",
        goodwill: "59596000000",
        investments: "None",
        longTermInvestments: "None",
        shortTermInvestments: "3721000000",
        otherCurrentAssets: "2581000000",
        otherNonCurrentAssets: "None",
        totalLiabilities: "106165000000",
        totalCurrentLiabilities: "30606000000",
        currentAccountsPayable: "3342000000",
        deferredRevenue: "15002000000",
        currentDebt: "12814000000",
        shortTermDebt: "6414000000",
        totalNonCurrentLiabilities: "75560000000",
        capitalLeaseObligations: "3283000000",
        longTermDebt: "50664000000",
        currentLongTermDebt: "6400000000",
        longTermDebtNoncurrent: "48828000000",
        shortLongTermDebtTotal: "84575000000",
        otherCurrentLiabilities: "8126000000",
        otherNonCurrentLiabilities: "12081000000",
        totalShareholderEquity: "23081000000",
        treasuryStock: "169640000000",
        retainedEarnings: "149506000000",
        commonStock: "59313000000",
        commonStockSharesOutstanding: "912800000",
      },
      {
        fiscalDateEnding: "2023-06-30",
        reportedCurrency: "USD",
        totalAssets: "132213000000",
        totalCurrentAssets: "34458000000",
        cashAndCashEquivalentsAtCarryingValue: "9394000000",
        cashAndShortTermInvestments: "9394000000",
        inventory: "1501000000",
        currentNetReceivables: "6368000000",
        totalNonCurrentAssets: "96170000000",
        propertyPlantEquipment: "5443000000",
        accumulatedDepreciationAmortizationPPE: "13145000000",
        intangibleAssets: "66881000000",
        intangibleAssetsExcludingGoodwill: "10496000000",
        goodwill: "56385000000",
        investments: "None",
        longTermInvestments: "1585000000",
        shortTermInvestments: "6904000000",
        otherCurrentAssets: "2731000000",
        otherNonCurrentAssets: "None",
        totalLiabilities: "109942000000",
        totalCurrentLiabilities: "32513000000",
        currentAccountsPayable: "3732000000",
        deferredRevenue: "15976000000",
        currentDebt: "13565000000",
        shortTermDebt: "6785000000",
        totalNonCurrentLiabilities: "77429000000",
        capitalLeaseObligations: "2828000000",
        longTermDebt: "54030000000",
        currentLongTermDebt: "6780000000",
        longTermDebtNoncurrent: "50691000000",
        shortLongTermDebtTotal: "87088000000",
        otherCurrentLiabilities: "8442000000",
        otherNonCurrentLiabilities: "12103000000",
        totalShareholderEquity: "22201000000",
        treasuryStock: "169581000000",
        retainedEarnings: "149318000000",
        commonStock: "58963000000",
        commonStockSharesOutstanding: "909855943",
      },
      {
        fiscalDateEnding: "2023-03-31",
        reportedCurrency: "USD",
        totalAssets: "133637000000",
        totalCurrentAssets: "35982000000",
        cashAndCashEquivalentsAtCarryingValue: "9337000000",
        cashAndShortTermInvestments: "9337000000",
        inventory: "1603000000",
        currentNetReceivables: "12626000000",
        totalNonCurrentAssets: "96048000000",
        propertyPlantEquipment: "5344000000",
        accumulatedDepreciationAmortizationPPE: "13417000000",
        intangibleAssets: "67098000000",
        intangibleAssetsExcludingGoodwill: "10905000000",
        goodwill: "56193000000",
        investments: "None",
        longTermInvestments: "1607000000",
        shortTermInvestments: "8057000000",
        otherCurrentAssets: "2501000000",
        otherNonCurrentAssets: "None",
        totalLiabilities: "111964000000",
        totalCurrentLiabilities: "30993000000",
        currentAccountsPayable: "3728000000",
        deferredRevenue: "16663000000",
        currentDebt: "9768000000",
        shortTermDebt: "4887000000",
        totalNonCurrentLiabilities: "80971000000",
        capitalLeaseObligations: "2963000000",
        longTermDebt: "55692000000",
        currentLongTermDebt: "4881000000",
        longTermDebtNoncurrent: "53826000000",
        shortLongTermDebtTotal: "86257000000",
        otherCurrentLiabilities: "8289000000",
        otherNonCurrentLiabilities: "12099000000",
        totalShareholderEquity: "21604000000",
        treasuryStock: "169544000000",
        retainedEarnings: "149253000000",
        commonStock: "58675000000",
        commonStockSharesOutstanding: "907500000",
      },
      {
        fiscalDateEnding: "2022-12-31",
        reportedCurrency: "USD",
        totalAssets: "127243000000",
        totalCurrentAssets: "29118000000",
        cashAndCashEquivalentsAtCarryingValue: "7886000000",
        cashAndShortTermInvestments: "7886000000",
        inventory: "1552000000",
        currentNetReceivables: "14209000000",
        totalNonCurrentAssets: "96874000000",
        propertyPlantEquipment: "5334000000",
        accumulatedDepreciationAmortizationPPE: "13361000000",
        intangibleAssets: "67133000000",
        intangibleAssetsExcludingGoodwill: "11184000000",
        goodwill: "55949000000",
        investments: "None",
        longTermInvestments: "142000000",
        shortTermInvestments: "852000000",
        otherCurrentAssets: "2610000000",
        otherNonCurrentAssets: "None",
        totalLiabilities: "105222000000",
        totalCurrentLiabilities: "31505000000",
        currentAccountsPayable: "4051000000",
        deferredRevenue: "15531000000",
        currentDebt: "9511000000",
        shortTermDebt: "4760000000",
        totalNonCurrentLiabilities: "83414000000",
        capitalLeaseObligations: "164000000",
        longTermDebt: "47190000000",
        currentLongTermDebt: "4676000000",
        longTermDebtNoncurrent: "46189000000",
        shortLongTermDebtTotal: "107759000000",
        otherCurrentLiabilities: "9788000000",
        otherNonCurrentLiabilities: "12243000000",
        totalShareholderEquity: "21944000000",
        treasuryStock: "169484000000",
        retainedEarnings: "149825000000",
        commonStock: "58343000000",
        commonStockSharesOutstanding: "906091977",
      },
    ],
  };
};

export const mockIncomeStatementResponse = () => {
  return {
    symbol: "IBM",
    annualReports: [
      {
        fiscalDateEnding: "2022-12-31",
        reportedCurrency: "USD",
        grossProfit: "32687000000",
        totalRevenue: "60530000000",
        costOfRevenue: "27842000000",
        costofGoodsAndServicesSold: "385000000",
        operatingIncome: "6408000000",
        sellingGeneralAndAdministrative: "18609000000",
        researchAndDevelopment: "6567000000",
        operatingExpenses: "26279000000",
        investmentIncomeNet: "None",
        netInterestIncome: "-1216000000",
        interestIncome: "162000000",
        interestExpense: "1216000000",
        nonInterestIncome: "365000000",
        otherNonOperatingIncome: "443000000",
        depreciation: "2407000000",
        depreciationAndAmortization: "2395000000",
        incomeBeforeTax: "1013000000",
        incomeTaxExpense: "-626000000",
        interestAndDebtExpense: "1216000000",
        netIncomeFromContinuingOperations: "1783000000",
        comprehensiveIncomeNetOfTax: "8134000000",
        ebit: "2229000000",
        ebitda: "4624000000",
        netIncome: "1639000000",
      },
      {
        fiscalDateEnding: "2021-12-31",
        reportedCurrency: "USD",
        grossProfit: "31486000000",
        totalRevenue: "57350000000",
        costOfRevenue: "25865000000",
        costofGoodsAndServicesSold: "300000000",
        operatingIncome: "4786000000",
        sellingGeneralAndAdministrative: "18745000000",
        researchAndDevelopment: "6488000000",
        operatingExpenses: "26700000000",
        investmentIncomeNet: "None",
        netInterestIncome: "-1155000000",
        interestIncome: "52000000",
        interestExpense: "1155000000",
        nonInterestIncome: "337000000",
        otherNonOperatingIncome: "225000000",
        depreciation: "3888000000",
        depreciationAndAmortization: "2529000000",
        incomeBeforeTax: "5867000000",
        incomeTaxExpense: "124000000",
        interestAndDebtExpense: "1155000000",
        netIncomeFromContinuingOperations: "4712000000",
        comprehensiveIncomeNetOfTax: "10582000000",
        ebit: "7022000000",
        ebitda: "9551000000",
        netIncome: "5743000000",
      },
      {
        fiscalDateEnding: "2020-12-31",
        reportedCurrency: "USD",
        grossProfit: "30865000000",
        totalRevenue: "55179000000",
        costOfRevenue: "24314000000",
        costofGoodsAndServicesSold: "439000000",
        operatingIncome: "4609000000",
        sellingGeneralAndAdministrative: "20561000000",
        researchAndDevelopment: "6262000000",
        operatingExpenses: "30966000000",
        investmentIncomeNet: "None",
        netInterestIncome: "-1288000000",
        interestIncome: "105000000",
        interestExpense: "1288000000",
        nonInterestIncome: "-92000000",
        otherNonOperatingIncome: "156000000",
        depreciation: "4227000000",
        depreciationAndAmortization: "2468000000",
        incomeBeforeTax: "4230000000",
        incomeTaxExpense: "-1360000000",
        interestAndDebtExpense: "1288000000",
        netIncomeFromContinuingOperations: "3932000000",
        comprehensiveIncomeNetOfTax: "4850000000",
        ebit: "5518000000",
        ebitda: "7986000000",
        netIncome: "5590000000",
      },
    ],
    quarterlyReports: [
      {
        fiscalDateEnding: "2023-09-30",
        reportedCurrency: "USD",
        grossProfit: "8023000000",
        totalRevenue: "14752000000",
        costOfRevenue: "6729000000",
        costofGoodsAndServicesSold: "42000000",
        operatingIncome: "1994000000",
        sellingGeneralAndAdministrative: "4458000000",
        researchAndDevelopment: "1685000000",
        operatingExpenses: "6029000000",
        investmentIncomeNet: "None",
        netInterestIncome: "-412000000",
        interestIncome: "None",
        interestExpense: "412000000",
        nonInterestIncome: "265000000",
        otherNonOperatingIncome: "97000000",
        depreciation: "521000000",
        depreciationAndAmortization: "572000000",
        incomeBeforeTax: "1863000000",
        incomeTaxExpense: "159000000",
        interestAndDebtExpense: "412000000",
        netIncomeFromContinuingOperations: "1714000000",
        comprehensiveIncomeNetOfTax: "2105000000",
        ebit: "2275000000",
        ebitda: "2847000000",
        netIncome: "1704000000",
      },
      {
        fiscalDateEnding: "2023-06-30",
        reportedCurrency: "USD",
        grossProfit: "8501000000",
        totalRevenue: "15475000000",
        costOfRevenue: "6974000000",
        costofGoodsAndServicesSold: "61000000",
        operatingIncome: "1984000000",
        sellingGeneralAndAdministrative: "4900000000",
        researchAndDevelopment: "1687000000",
        operatingExpenses: "6517000000",
        investmentIncomeNet: "None",
        netInterestIncome: "-423000000",
        interestIncome: "201000000",
        interestExpense: "423000000",
        nonInterestIncome: "163000000",
        otherNonOperatingIncome: "39000000",
        depreciation: "520000000",
        depreciationAndAmortization: "557000000",
        incomeBeforeTax: "2002000000",
        incomeTaxExpense: "419000000",
        interestAndDebtExpense: "423000000",
        netIncomeFromContinuingOperations: "1581000000",
        comprehensiveIncomeNetOfTax: "1863000000",
        ebit: "2425000000",
        ebitda: "2982000000",
        netIncome: "1583000000",
      },
      {
        fiscalDateEnding: "2023-03-31",
        reportedCurrency: "USD",
        grossProfit: "7509000000",
        totalRevenue: "14252000000",
        costOfRevenue: "6743000000",
        costofGoodsAndServicesSold: "30000000",
        operatingIncome: "256000000",
        sellingGeneralAndAdministrative: "4853000000",
        researchAndDevelopment: "1655000000",
        operatingExpenses: "7253000000",
        investmentIncomeNet: "None",
        netInterestIncome: "-367000000",
        interestIncome: "170000000",
        interestExpense: "367000000",
        nonInterestIncome: "-93000000",
        otherNonOperatingIncome: "22000000",
        depreciation: "527000000",
        depreciationAndAmortization: "547000000",
        incomeBeforeTax: "1051000000",
        incomeTaxExpense: "124000000",
        interestAndDebtExpense: "367000000",
        netIncomeFromContinuingOperations: "934000000",
        comprehensiveIncomeNetOfTax: "888000000",
        ebit: "1418000000",
        ebitda: "1965000000",
        netIncome: "927000000",
      },
      {
        fiscalDateEnding: "2022-12-31",
        reportedCurrency: "USD",
        grossProfit: "9632000000",
        totalRevenue: "16690000000",
        costOfRevenue: "7058000000",
        costofGoodsAndServicesSold: "190000000",
        operatingIncome: "3294000000",
        sellingGeneralAndAdministrative: "4766000000",
        researchAndDevelopment: "1604000000",
        operatingExpenses: "6338000000",
        investmentIncomeNet: "None",
        netInterestIncome: "-313000000",
        interestIncome: "64000000",
        interestExpense: "313000000",
        nonInterestIncome: "-394000000",
        otherNonOperatingIncome: "36000000",
        depreciation: "570000000",
        depreciationAndAmortization: "567000000",
        incomeBeforeTax: "3154000000",
        incomeTaxExpense: "444000000",
        interestAndDebtExpense: "313000000",
        netIncomeFromContinuingOperations: "2870000000",
        comprehensiveIncomeNetOfTax: "3109000000",
        ebit: "3467000000",
        ebitda: "4034000000",
        netIncome: "2710000000",
      },
    ],
  };
};

export const mockDailyData = () => {
  return {
    "Meta Data": {
      "1. Information": "Daily Prices (open, high, low, close) and Volumes",
      "2. Symbol": "IBM",
      "3. Last Refreshed": "2023-11-02",
      "4. Output Size": "Compact",
      "5. Time Zone": "US/Eastern",
    },
    "Time Series (Daily)": {
      "2023-11-02": {
        "1. open": "145.7700",
        "2. high": "147.1000",
        "3. low": "144.8400",
        "4. close": "147.0100",
        "5. volume": "3902657",
      },
      "2023-11-01": {
        "1. open": "145.0000",
        "2. high": "146.5100",
        "3. low": "144.4500",
        "4. close": "145.4000",
        "5. volume": "4750081",
      },
      "2023-11-03": {
        "1. open": "145.7700",
        "2. high": "147.1000",
        "3. low": "144.8400",
        "4. close": "147.0100",
        "5. volume": "3902657",
      },
      "2023-11-04": {
        "1. open": "145.7700",
        "2. high": "147.1000",
        "3. low": "144.8400",
        "4. close": "147.0100",
        "5. volume": "3902657",
      },
    },
  };
};
