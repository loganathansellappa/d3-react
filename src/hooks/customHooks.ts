import axios from "axios";
import { useQuery } from "react-query";
import {
  NormalizedCompanySchema,
  BalanceSheetSchema,
  IncomeStatementSchema,
  ApiLimitSchema,
  TimeSeriesDailySchema,
} from "../@types/zodSchema";
import { z } from "zod";
import {serverData, transformTimeSeriesDaily} from "../utils/HelperUtils";

const TransformedTimeSeriesDailySchema = TimeSeriesDailySchema.transform(
  (data) => transformTimeSeriesDaily(data["Time Series (Daily)"]),
);
const fetchData = async (fn: string = "OVERVIEW") => {
  const url = `${serverData().baseUrl}?function=${fn}&symbol=IBM&apikey=${
    serverData().apiKey
  }`;
  const response = await axios.get(url);
  const schemaMap: Record<string, z.Schema<unknown>> = {
    OVERVIEW: NormalizedCompanySchema,
    BALANCE_SHEET: BalanceSheetSchema,
    INCOME_STATEMENT: IncomeStatementSchema,
    TIME_SERIES_DAILY: TransformedTimeSeriesDailySchema,
    LIMIT_EXCEEDED: ApiLimitSchema,
    // Add more functions and schemas as needed
  };
  if (response.data.Information) {
    throw new Error(response.data.Information);
  }
  return schemaMap[fn].parse(response.data);
};

export function useApi(fn: string = "OVERVIEW") {
  return useQuery({
    queryKey: fn,
    queryFn: () => fetchData(fn),
  });
}


