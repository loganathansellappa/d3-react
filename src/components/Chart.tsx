import React, { useCallback, useMemo, useState } from "react";
import { useApi } from "../hooks/useOverview";
import LoadingSpinner from "./LoadingSpinner";
import * as d3 from "d3";
import { AreaChart } from "./charts/AreaChart";
import { CandleChart } from "./charts/CandleChart";
import "./Chart.scss";
import { ChartDatum, ChartType } from "../@types/ChartData";

export const Chart: React.FC = () => {
  const { data, isLoading, isError, error } = useApi("TIME_SERIES_DAILY");
  const [chartType, setChartType] = useState<ChartType>("AREA_STICK");

  const tableData = useMemo((): Array<ChartDatum> => {
    if (!data) {
      return [];
    }
    const parseDate = d3.timeParse("%Y-%m-%d");

    const stockData = Object.entries(data)
      .map(([date, values]) => ({
        ...values,
        Date: parseDate(date),
      }))
      .reverse();
    return stockData;
  }, [data]);

  const getChart = useCallback(() => {
    switch (chartType) {
      case "CANDLE_STICK":
        return <CandleChart data={tableData} />;
      default:
        return <AreaChart data={tableData} />;
    }
  }, [tableData, chartType]);

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (isError && error instanceof Error) {
    return <div>Error fetching data {error.message}</div>;
  }

  return (
    <div>
      <div className={"button-container"}>
        <button
          className={"button"}
          onClick={() => setChartType("CANDLE_STICK")}
        >
          CANDLE STICK
        </button>
        <button className={"button"} onClick={() => setChartType("AREA_STICK")}>
          AREA STICK
        </button>
      </div>
      {getChart()}
    </div>
  );
};
