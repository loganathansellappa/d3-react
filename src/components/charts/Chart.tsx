import React, { useCallback, useMemo, useState } from "react";
import { useApi } from "../../hooks/customHooks";
import LoadingSpinner from "../loader/LoadingSpinner";
import * as d3 from "d3";
import { AreaChart } from "./AreaChart";
import { CandleChart } from "./CandleChart";
import "./Chart.scss";
import {
  ChartComponentProps,
  ChartDatum,
  ChartType,
} from "../../@types/ChartData";

interface ChartProps extends Omit<ChartComponentProps, "data"> {}
export const Chart: React.FC<ChartProps> = ({ cwidth, cheight }) => {
  const { data, isLoading, isError, error } = useApi("TIME_SERIES_DAILY");
  const [chartType, setChartType] = useState<ChartType>("AREA_STICK");

  const tableData = useMemo((): Array<ChartDatum> => {
    if (!data) {
      return [];
    }
    const parseDate = d3.timeParse("%Y-%m-%d");

    const stockData = Object.entries(data)
      .map(([date, values]) => ({
        ...values,
        Date: parseDate(date),
      }))
      .reverse();
    return stockData;
  }, [data]);

  const getChart = useCallback(() => {
    switch (chartType) {
      case "CANDLE_STICK":
        return (
          <CandleChart cwidth={cwidth} cheight={cheight} data={tableData} />
        );
      default:
        return <AreaChart cwidth={cwidth} cheight={cheight} data={tableData} />;
    }
  }, [tableData, chartType, cwidth, cheight]);

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (isError && error instanceof Error) {
    return <div>Error fetching data {error.message}</div>;
  }

  return (
    <div className={'chart-wrapper'}>
      <div className={"button-container"}>
        <button
          className={"button"}
          onClick={() => setChartType("CANDLE_STICK")}
        >
          CANDLE STICK
        </button>
        <button className={"button"} onClick={() => setChartType("AREA_STICK")}>
          AREA STICK
        </button>
      </div>

      {getChart()}
    </div>
  );
};
