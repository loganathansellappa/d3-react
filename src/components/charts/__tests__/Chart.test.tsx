import { QueryClient, QueryClientProvider, UseQueryResult } from "react-query";
import { mockDailyData, renderWithRouter } from "../../../utils/test";
import * as Hooks from "../../../hooks/customHooks";
import { Chart } from "../Chart";
import { fireEvent } from "@testing-library/react";

describe("Chart", () => {
  let element: { container: HTMLElement };
  beforeEach(() => {
    jest.spyOn(Hooks, "useApi").mockReturnValue({
      data: mockDailyData(),
      isError: false,
      isLoading: false,
      refetch: jest.fn(),
    } as unknown as UseQueryResult<unknown, unknown>);
    const queryClient = new QueryClient();
    element = renderWithRouter(
      <QueryClientProvider client={queryClient}>
        <Chart />
      </QueryClientProvider>,
    );
  });

  test("renders Overview component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });

  test("displays area chart", () => {
    expect(element.container.textContent).toContain("Area");
  });

  test("displays area chart by default", () => {
    const buttonOne =
      element.container.getElementsByClassName("button-active")[0];
    expect(buttonOne.textContent).toContain("Area");
  });

  test("displays candle chart when clicking button", () => {
    const button = element.container.getElementsByClassName("button")[1];
    fireEvent.click(button);
    const buttonTwo = element.container.getElementsByClassName("button")[1];
    expect(buttonTwo.textContent).toContain("Candle");
  });

  test("displays svg", () => {
    const chart = element.container.getElementsByTagName("svg")[0];
    expect(chart).toBeTruthy();
  });
});
