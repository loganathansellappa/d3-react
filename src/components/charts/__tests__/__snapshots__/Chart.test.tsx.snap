// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart renders Overview component 1`] = `
<div
  class="chart-wrapper"
>
  <h2
    class="title"
  >
    Daily Data
  </h2>
  <div
    class="button-container"
  >
    <button
      class="button button-active"
    >
      Area Chart
    </button>
    <button
      class="button "
    >
      Candle Chart
    </button>
  </div>
  <div
    class="chart-container"
  >
    <svg
      viewBox="0 0 1600 800"
    >
      <g
        transform="translate(10,100)"
      />
      <g
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="middle"
        transform="translate(0,690)"
      >
        <path
          class="domain"
          d="M0.5,6V0.5H1530.5V6"
          stroke="currentColor"
        />
      </g>
      <g
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="start"
        transform="translate(1530,0)"
      >
        <path
          class="domain"
          d="M6,690.5H0.5V0.5H6"
          stroke="currentColor"
        />
      </g>
      <path
        class="area"
        d="MNaN,NaNLNaN,NaNLNaN,690LNaN,690Z"
        style="fill: #007bff; opacity: 0.5;"
      />
      <path
        class="line"
        d="MNaN,NaNLNaN,NaN"
        fill="none"
        stroke="#007bff"
        stroke-width="1"
      />
      <circle
        fill="red"
        opacity="0.7"
        r="0"
        style="stroke: white; pointer-events: none;"
      />
      <line
        class="crosshair-x"
        id="crosshair-x"
        stroke="gray"
        stroke-dasharray="3,3"
      />
      <line
        class="crosshair-y"
        id="crosshair-y"
        stroke="gray"
        stroke-dasharray="3,3"
      />
      <rect
        class="background-area-rect"
        height="690"
        width="1530"
      />
    </svg>
    <div
      class="tooltip"
    />
    <div
      class="tooltip"
    />
  </div>
</div>
`;
